src/esi/Expression.cc

```diff
@@ -215,6 +228,7 @@ evalnegate(stackmember * stack, int *depth, int whereAmI, stackmember * candidat
     /* copy down */
     --(*depth);
 
+    Must2(*depth >= 0 && *depth < STACK_DEPTH_LIMIT && whereAmI >= 0 && whereAmI < STACK_DEPTH_LIMIT, "negate expression too complex");
```

Comment:
The last component of this check feels too late and too lax because we have already accessed `stack[whereAmI + 1]` a few lines above. It also feels unnecessary because `whereAmI` is `*depth - 2` and depth has been validated by the earlier components.

Reply:
Fixed, there are now two Must, one for each parameter checked.

Reply:
> Fixed, there are now two Must

The conditions have changed, but the same rule of thumb applies to the remaining ones: One Must() or assert() per ANDed condition. It may especially be useful here, where one condition is driven by the immediate user input while the other one should probably never fail (or, at the very least, is violated by ESIs processed earlier). For example:

```
    Must2(*depth >= 0, "negate expression too complex");
    Must2(*depth < ESI_STACK_DEPTH_LIMIT, ...); // too big even after we decremented it!
```